### PR TITLE
Add utils methods in pinecone provider

### PR DESCRIPTION
**Pull request by utkarsharma2**
Added the following util methods in the pinecone provider that are useful during the development phase of the LLM pipeline and are intended to be used from within the Python operator. Using these methods we can leverage the airflow's connections and can save time by avoiding writing boilerplate code from scratch. 
1. list_indexes
2. create_index
3. describe_index
4. configure_index
5. create_collection
6. delete_collection
7. list_collections
8. query_vector
9. upsert_data_async
10. describe_index_stats

### CODE REVIEW

File: airflow/providers/pinecone/hooks/pinecone.py [hunk 1]
```diff
@@ -126,3 +133,212 @@ def upsert(
...
+        timeout: int | None = None,
+        index_type: str | None = "approximated",
+        metric: str | None = "cosine",
+        replicas: int | None = 1,
+        shards: int | None = 1,
+        pods: int | None = 1,
+        pod_type: str | None = "p1",
+        index_config: dict[str, str] | None = None,
+        metadata_config: dict[str, str] | None = None,
+        source_collection: str | None = "",
```
Comment (pankajastro) [inline airflow/providers/pinecone/hooks/pinecone.py]:
```suggestion
        index_type: str | None = "approximated",
        metric: str | None = "cosine",
        replicas: int | None = 1,
        shards: int | None = 1,
        pods: int | None = 1,
        pod_type: str | None = "p1",
        index_config: dict[str, str] | None = None,
        metadata_config: dict[str, str] | None = None,
        source_collection: str | None = "",
        timeout: int | None = None,
```

nit to keep order same as docstring

File: airflow/providers/pinecone/hooks/pinecone.py [hunk 2]
```diff
@@ -126,3 +133,212 @@ def upsert(
...
+    ) -> None | list[Any]:
+        """
+        Upserts (insert/update) data into the Pinecone index.
+
+        :param index_name: Name of the index.
+        :param data: List of tuples to be upserted. Each tuple is of form (id, vector, metadata).
+                     Metadata is optional.
+        :param async_req: If True, upsert operations will be asynchronous.
+        :param pool_threads: Number of threads for parallel upserting. If async_req is True, this must be provided.
+        :return: Upsert responses or None for async operations.
```
Comment (pankajastro) [inline airflow/providers/pinecone/hooks/pinecone.py]:
```suggestion
```

maybe if want to stay consistent?

**Comment by kaxil** [comment]
Test failed:

```

=========================== short test summary info ============================
FAILED tests/integration/executors/test_celery_executor.py::TestCeleryExecutor::test_celery_integration[redis://redis:6379/0] - AssertionError: assert equals failed
  [                                [                               
    (                                (                             
      'success',                       'success',                  
      'fake_simple_ti',                'fake_simple_ti',           
      datetime.datetime(2023, 11,      datetime.datetime(2023, 11, 
   7, 12, 58, 35, 619240),          7, 12, 58, 35, 619240),        
      0,                               0,                          
    ),                               ),                            
                                     (                             
                                       'fail',                     
                                       'fake_simple_ti',           
                                       datetime.datetime(2023, 11, 
                                    7, 12, 58, 35, 619240),        
                                       0,                          
                                     ),                            
  ]                                ]
======== 1 failed, 14 passed, 40 skipped, 4 warnings, 3 rerun in 14.59s ========
```

**Comment by kaxil** [comment]
All the tests in [DB:Postgres12,Py3.8: Always Providers[pinecone]](https://github.com/apache/airflow/actions/runs/6784692186/job/18441677999?pr=35502#logs) were skipped, expected?:

**Comment by utkarsharma2** [comment]
@kaxil I guess they should be executed here - https://github.com/apache/airflow/actions/runs/6784692186/job/18441680475?pr=35502

**Comment by utkarsharma2** [comment]
> Test failed (https://github.com/apache/airflow/actions/runs/6784692186/job/18441679074?pr=35502), although this looks unrelated
> 
> ```
> 
> =========================== short test summary info ============================
> FAILED tests/integration/executors/test_celery_executor.py::TestCeleryExecutor::test_celery_integration[redis://redis:6379/0] - AssertionError: assert equals failed
>   [                                [                               
>     (                                (                             
>       'success',                       'success',                  
>       'fake_simple_ti',                'fake_simple_ti',           
>       datetime.datetime(2023, 11,      datetime.datetime(2023, 11, 
>    7, 12, 58, 35, 619240),          7, 12, 58, 35, 619240),        
>       0,                               0,                          
>     ),                               ),                            
>                                      (                             
>                                        'fail',                     
>                                        'fake_simple_ti',           
>                                        datetime.datetime(2023, 11, 
>                                     7, 12, 58, 35, 619240),        
>                                        0,                          
>                                      ),                            
>   ]                                ]
> ======== 1 failed, 14 passed, 40 skipped, 4 warnings, 3 rerun in 14.59s ========
> ```

This indeed is unrelated.

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
> @kaxil I guess they should be executed here - https://github.com/apache/airflow/actions/runs/6784692186/job/18441680475?pr=35502 Although I'm not seeing the list of tests ran as part of the job.

Verified they are running:

- https://github.com/apache/airflow/actions/runs/6784692186/job/18441680475?pr=35502#step:5:618
- https://github.com/apache/airflow/actions/runs/6784692186/job/18441680475?pr=35502#step:5:718

**Review by pankajastro — APPROVED** [review]

**Pull request merged**